Restructure dashboard

lboeman opened the pull request:
Will need a Rebase on #31 

Changes summary:
- Reorganize components directory, with generic/helper components in the root and a `model` directory that contains all of the vue components for editing objects found in `src/types`.
- Update components to use new api fields.
- Create `Home.vue` view component that contains header, footer, nav, and a body element that renders the current route's component. Nav displays differently for authenticated/unauthenticated users.

A large portion of this PR is the unfortunate input element pattern like:
```
 <b>Make and Model:</b>
  <input v-model="pvarray.make_model" />
  <help :helpText="this.definitions.properties.make_model.description" />
  <br />
  <span style="color:#F00;" v-if="'make_model' in this.errors">
    {{ this.errors.make_model }}
    <br />
  </span>
```
I haven't found a good way to factor out this pattern that isn't just as verbose and more obscure so I've left it for now.

alorenzo175 commented:
> A large portion of this PR is the unfortunate input element pattern like:
> 
> ```
>  <b>Make and Model:</b>
>   <input v-model="pvarray.make_model" />
>   <help :helpText="this.definitions.properties.make_model.description" />
>   <br />
>   <span style="color:#F00;" v-if="'make_model' in this.errors">
>     {{ this.errors.make_model }}
>     <br />
>   </span>
> ```
> 
> I haven't found a good way to factor out this pattern that isn't just as verbose and more obscure so I've left it for now.

Sorry I didn't really give this much thought before, but it's repeated so often that it'd be easy to miss something now or if something needs to change later. So, even if it's about as verbose, is it possible to make a component like this?
```
<parameter-field :title="Make and Model" :model="pvarray.make_model" :key="make_model"></parameter-field>
```

alorenzo175 commented on dashboard/src/components/model/Inverter.vue:
```diff
@@ -24,30 +39,39 @@
 Vue.component("arrays-view", ArraysView);
 Vue.component("inverter-parameters", InverterParametersView);
+Vue.component("loss-parameters", LossParametersView);
+Vue.component("help", HelpPopup);
```
Should all of these components be registered globally or at a higher level instead of in each component?

https://vuejs.org/v2/guide/components-registration.html#Global-Registration

alorenzo175 commented on dashboard/src/components/model/LossParameters.vue:
```diff
@@ -0,0 +1,135 @@
+        {{ this.errors.snow }}
+        <br />
+      </span>
+      <b>mismath:</b>
```
```suggestion
      <b>mismatch:</b>
```

alorenzo175 commented on dashboard/src/views/SystemSpec.vue:
```diff
@@ -0,0 +1,218 @@
+    }
+  }
+
+  async loadSystem() {
```
shouldn't the system already be in the store? In any case, I think a separate interface to the API detached from views or components might be helpful. Maybe that is a future issue

alorenzo175 commented on dashboard/src/views/SystemSpec.vue:
```diff
@@ -0,0 +1,218 @@
+      })
+    });
+    if (response.ok) {
+      const system: System = await response.json();
```
this is really a StoredSystem. So for now the system should be the .defnition attribute of the returned json

alorenzo175 commented on dashboard/src/components/model/ModuleParameters.vue:
```diff
@@ -0,0 +1,205 @@
+      <br />
+    </div>
+    <div v-if="model == 'pvsyst'">
+      <b>gamma_ref:</b>
+      <input type="number" v-model.number="parameters.gamma_ref" />
+      <help :helpText="this.definitions.properties.gamma_ref.description" />
+      <br />
+      <span style="color:#F00;" v-if="'gamma_ref' in this.errors">
+        {{ this.errors.gamma_ref }}
+        <br />
+      </span>
+      <b>mu_gamma:</b>
+      <input v-model="parameters.mu_gamma" />
+      <help :helpText="this.definitions.properties.mu_gamma.description" />
+      <br />
+      <span style="color:#F00;" v-if="'mu_gamma' in this.errors">
+        {{ this.errors.mu_gamma }}
+        <br />
+      </span>
```
Why does gamma_ref have `type="number" v-model.number=` but mu_gamma doesn't?

alorenzo175 replied:
maybe this is causing one validation issue (along with the added example in the openapi spec)

![image](https://user-images.githubusercontent.com/4625457/101401779-0a0cf700-3890-11eb-82b2-c3db14ad4e7b.png)

alorenzo175 commented on dashboard/src/store/actions.ts:
```diff
@@ -1,15 +1,19 @@
-    context.commit("updateSystemsList", systemlist as Array<System>);
+    const systemMapping: Record<string, System> = {};
+    for (const system of systemlist) {
+      systemMapping[system.system_id] = system as System;
```
api has changed so this should be 
```suggestion
      systemMapping[system.object_id] = system.definition as System;
```
or whatever for typescript etc

lboeman commented on dashboard/src/components/ModelField.vue:
```diff
@@ -0,0 +1,32 @@
+<template>
+  <div class="model-field">
+    <b>{{ title }}:</b>
+    <input v-if="inputType == 'number'" type="number" v-model.number="$parent.parameters[fieldName]" />
+    <input v-if="inputType != 'number'" v-model="$parent.parameters[fieldName]" />
+    <help :helpText="$parent.definitions.properties[fieldName].description" />
+    <br />
+    <span style="color: #F00;" v-if="fieldName in $parent.errors">
+      {{ $parent.errors[fieldName] }}
+    </span>
+  </div>
+</template>
+
+<script lang="ts">
+import { Component, Prop, Vue } from "vue-property-decorator";
+
+@Component
+export default class ModelField extends Vue{
+  @Prop() title!: string;
+  @Prop() inputType?: string;
+  @Prop() fieldName!: string;
+  //@Prop() parameters!: Record<string, any>;
+  //@Prop() errors!: Record<string, any>;
+  //@Prop() definitions!: Record<string, any>;
+
+}
+</script>
+
+<!-- Add "scoped" attribute to limit CSS to this component only -->
+<style scoped>
+
+</style>
```
@alorenzo175 I created this component to factor out the copy pasta. The way it's written here, it reads and alters the props of its parents directly. We could also use the commented out props and pass all of the props in. I think this should be okay since this will always be a leaf node with no children, and I can't think of any issues it may cause. Do you have any objections or concerns with this?

alorenzo175 replied:
I'm definitely just learning Vue, so how is this altering the parent?

Also, I don't think the errors span is working. For example, modifying models.py adding `ge=0` to `u_c` and then entering -1 for uC does not generate 'should be >= 0'

lboeman replied:
Ah, thanks. there was no validate function for TemperatureParameters. errors should work now, an empty field will yield "should be a number".

The `v-model` bindings are accessing the `$parent` directly. Essentially, anywhere this is used is agreeing to have it's `parameters` mutated by this component. It's not too far from what the rest of these components are doing, e.g. clicking "add inverter" appends a new inverter object to the systems `inverters` property, but hands it off to a component to be altered.

To explain why I'm double checking, All of the model components break from the suggested Vue pattern of passing props down through the hierarchy, and emitting events up to mutate data. But since these components are meant to mutate the system specification, which will only exist in memory on the page, it seems like a logical way to go.

alorenzo175 replied:
I see. I guess that's OK. good to go change the rest to use this component

alorenzo175 commented:
Looks good, but on Firefox + fedora only integers are allowed in the "number" fields. See toaster photo

![PXL_20201208_181713963.jpg](https://user-images.githubusercontent.com/4625457/101524403-1d2ece00-3947-11eb-9d35-5315635a4682.jpg)

alorenzo175 commented:
Apparently not a problem in midori, but is a problem in chrome with the hover showing up but no red outline

alorenzo175 commented:
ah, may need to set the `step="0.01"` or similar https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number.

lboeman commented:
> ah, may need to set the `step="any"` https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number.
> 
> ~It seems that you need to set the step to the smallest decimal place allowed to avoid the warning~

Thanks, added. I also realized I was missing the `strings` and `modules_per_string` field, which have a type `integer` in the spec, so I've added that.

lboeman commented:
This does have a bug that causes errors with similarly named fields to trigger errors upstream in the model. For example, setting an array name to "name*" triggers errors on inverter names and the system name. This is caused by a naive lookup of the field name in the errors object. But, I think there's probably a better generic way to do it that will work with API errors as well, so I'll hold off for another PR to fix that particular bug.

alorenzo175 commented:
> This does have a bug that causes errors with similarly named fields to trigger errors upstream in the model. For example, setting an array name to "name*" triggers errors on inverter names and the system name. This is caused by a naive lookup of the field name in the errors object. But, I think there's probably a better generic way to do it that will work with API errors as well, so I'll hold off for another PR to fix that particular bug.

Ok, let's make an issue for that
